### ISSUE TITLE
Showing what file is being hashed when scanning

### ISSUE THREAD

**Issue by Avamander** [syncthing/syncthing#3399]
For the last few days when I tried to sync a 10GB folder with a lot of subfolders and different sized files the status of the folder was just "Scanning" I'd really appreciate it when it'd actually show me if there was any progress made (showing hashed filename) or it's just stuck.

**Comment by calmh** [comment]
There should be a percentage indicator and time left approximation, with a hover showing the scan rate.

**Comment by Avamander** [comment]
Nope, I do not have that. Using v0.13.10, Linux (64 bit). It seems to appear when it reaches last files (shows 9 sec left for 10 minutes for ex). Unreliable with low hash speeds it seems?

**Comment by calmh** [comment]
No, that is the entire hashing phase then. The phase before showing a percentage is just looking for changes by checking modtimes, to determine what files to hash.

**Comment by calmh** [comment]
Sorry, I misread. How slow is this box anyway, that measurement is usually accurate almost to the second?

**Comment by AudriusButkevicius** [comment]
Scanning without a percentage means it's still walking the filesystem trying to understand how many files there are.

**Comment by Avamander** [comment]
There are about 177281 items sized 9.5GB total, single thread hash
performance is 80MB/s. It takes a lot of CPU and disk bandwidth for
syncthing to index it. That is one of the reasons I can't run Syncthing
just as I could Dropbox in the background.

>

**Comment by AudriusButkevicius** [comment]
It indexes data once, unless it changes.

**Comment by Avamander** [comment]
I don't exactly keep it constant either.

**Comment by AudriusButkevicius** [comment]
Well at the point actual scanning is happening, and if its happening for more than 5 seconds you'd get a progress bar.

If there is no progress bar, it potentially is just walking the filesystem, or the scan is just so short that it hasn't even got an announcement for progress yet.

Anyways, what you are asking for is already there, so I am not sure if there is a reason for this ticket to stay open.

**Comment by AudriusButkevicius** [comment]
Sorry, it's a lie, what you are after is knowing which file is being hashed, which is not available.

**Comment by calmh** [comment]
Well that but the thing that is strange is the "shows 9 sec left for 10 minutes". Something is broken in this setup, but I'm not sure what.

**Comment by Avamander** [comment]
Encountered this again with the latest version available from the repo. First of all, it marked almost all of the folder as out of sync (there is no way that much has changed). Secondly it has shown that 10 sec remaining for an hour, other folders sync fine meanwhile. Any ideas?

```
 Out of Sync Items	146438 items, ~9.12 GiB
 Scan Time Remaining	~ 0m 10s
```

**Comment by calmh** [comment]
There's something buggy with your UI. If it's showing out of sync items it's not scanning, and if it's not scanning there should be bo remaining time. Screenshots would help, and check the JavaScript console for errors.

**Comment by lkwg82** [comment]
@Avamander do you still have this issue, with a more recent release?

**Comment by Avamander** [comment]
I can't see it, but I see >100% scanned quite often (when files are continuously added to the folder for ex.).

**Comment by Tremelune** [comment]
I often see "Scanning" with no percentage for giant directories on network shares. It would be convenient to show some kind of progress indicator for the hashing step (which can easily take minutes).

I would assume that simply counting the number of files/dirs under a particular directory would be fast, such that you could at least know the total of the number you must walk to hash.

Something like `find . | wc` seems quite fast, though I realize it's not apples to apples.

**Comment by calmh** [comment]
The stage that does not have a progress indicator is the walk/stat stage, essentially equivalent to your find but with corresponding database lookups. (I’m not sure if find stats all files by default or gets the info it needs directly from the readdir.)

Once something to hash is found, you get the progress indicator. The assumption is that normally the walk stage should be quite fast, and we can’t know a priori how much there is to walk.

**Comment by Tremelune** [comment]
The assumption that the walk stage will be fast seems false for certain kinds of shares. In my case, a 5TB directory with 20k files on a wired Drobo 5D will show "Scanning" with no progress indicator for minutes. Maybe it's just my setup.

I guess the question is, how much overhead would be added by doing the fastest possible file/dir count before any other aspect of the walking was performed?

**Comment by AudriusButkevicius** [comment]
That's what the initial walk is doing.

**Comment by imsodin** [comment]
I think it isn't what the original issue was about, but I can see the merit of some kind of status indication during the walking phase in very slow setups. One way I see is to use the lexicographic order of walking, i.e. just send the currently walked path. I just don't have a good idea how to display that in the GUI. Displaying just a shortened path like "Scanning (firstdi...ilename)" (for e.g. `firstidir/seconddir/thirddir/filename`) might work, but I am not sure whether it is clear. Maybe give it an entirely different name, like "Preparing (firstdi...ilename)", but that's so generic. Any ideas?

**Comment by adamralph** [comment]
How about the filename (shortened if required)? For systems where files are read very quickly, there could be lower limit for the amount of time a given filename is displayed, e.g. 500ms. That could mean that the display of most of the filenames is skipped, but that shouldn't matter. It would provide the reassurance to the user that "something is happening", which I think was the original intent of the issue.

**Comment by imsodin** [comment]
That part isn't my problem. In the scanning implementation the only thing that makes sense (no more resources, reasonably easy) is the same system as with the current hashing progress events, i.e. send a filename/path every interval (default every 2s). I expect this to be just a bit of work, but not inherently difficult. For the usual case where walking is fast, there will be no event at all and thus direct display of the "Scanning (X%)" indication, which is fine. My question is how/where to display the walking information (filename) in the web UI very concretely, in a way that it's clear what it signifies, isn't too cluttery, ...

**Comment by Tremelune** [comment]
I think your suggestion would do the trick. Maybe just display a small, light, truncated path when a directory is taking longer than expected to walk:

http://ensaster.com/pics/github/sync.png

**Comment by AudriusButkevicius** [comment]
This would just fire thousands of events over a very short period of time and slow down then scan process even more.

**Comment by Tremelune** [comment]
Not if it...didn't. The update could be as simple as announcing what directory was currently being walked every second or so.

**Comment by calmh** [comment]
We already have the scanning progress events. When enabled, we could fire them in this phase too, on some interval. So, yeah.

**Comment by calmh** [comment]
So, eh, I missed that this was an existing issue about something else when I retitled it. But since the original one was unclear as to the cause, I think this is fine. Object if not.

**Comment by imsodin** [comment]
What I was thinking about on the non-GUI side is pretty simple: https://github.com/imsodin/syncthing/commit/619ba6b3cb7cda1f3a394b0bbd3cdc39884877ff  
If anyone wants to do the GUI part, feel free to use that if it suits your needs - I am not keen to work on  GUI.
